### PR TITLE
Feature/meeting query algo

### PR DESCRIPTION
**Pull request by jamohile**
**Work Done**
- Using test-driven dev, write algorithm to find potential meeting times given a set of mandatory/optional attendees and their pre-existing events.
- This approach contains a bit of duplication, since it divides the attendees into two classes: mandatory and optional. These two classes have almost identical logic, except for the fact that optional should only be considered if mandatory is fulfilled. 
 - I feel like there is a more elegant way to consider these two classes simultaneously
 - But for now, this works, and should serve as a good V1 until that can be found. 
 - Additionally all duplication is small-chunks, and located in very close proximity such that it doesn't take away from readability. As of now, abstracting things away to get rid of the duplication but keep the slight differences would probably be detrimental.

**ALL TESTS PASSING**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,180 @@
...
+    }
+  };
+
+  /** Finds potential meeting times */
+  public Collection<TimeRange> query(
+    Collection<Event> events,
+    MeetingRequest request
+  ) {
+    // Potential windows that only include mandatory attendees.
+    ArrayList<TimeRange> timeSlots = new ArrayList<>();
```
Comment (shijun-z) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: Interface is preferred when appropriate: 

List<TimeRange> timeSlots = new ArrayList<>();

Effective java:
If appropriate interface types exist, then parameters, return values, and fields should all be declared using interface types. If you get into the habit of using interface types, your program will be much more flexible. It is entirely appropriate to refer to an object by a class if no appropriate interface exists.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,180 @@
...
+      // If this window contains members from a certain class, update their respective block extent.
+      if (containsOptionalAttendees) { // We group all optional attendees as one lump for now.
+        optionalBlockExtent = Math.max(optionalBlockExtent, when.end());
+      }
+      if (containsMandatoryAttendees) {
+        mandatoryBlockExtent = Math.max(mandatoryBlockExtent, when.end());
+      }
+    }
+
+    // At this point there are no more events, so remaining might be a contiguous potential window.
```
Comment (shijun-z) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:133]: (nitpicking) Is it possible to add some tricks here so below code is no longer needed, e.g., add a empty event to the tail of the event list?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (shijun-z) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: Interface is preferred when appropriate:

List timeSlots = new ArrayList<>();

Effective java:
If appropriate interface types exist, then parameters, return values, and fields should all be declared using interface types. If you get into the habit of using interface types, your program will be much more flexible. It is entirely appropriate to refer to an object by a class if no appropriate interface exists.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -211,7 +278,7 @@ public void justEnoughRoom() {
     Collection<Event> events = Arrays.asList(
         new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
             Arrays.asList(PERSON_A)),
-        new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, true),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, false),
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:281]: Why the change?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -121,6 +122,72 @@ public void everyAttendeeIsConsidered() {
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:125]: Another test that's worth considering is one in which the events are explicitly out of order.  Your algorithm relies on the events being sorted (and it does sort them) but eyeballing it I don't see any tests that would fail if that sorting were removed.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,180 @@
...
+    }
+
+    // If possible, return the windows that include all optional. Otherwise, just return mandatory.
+    return timeSlotsWithOptional.size() > 0 ? timeSlotsWithOptional : timeSlots;
+  }
+
+  /**
+   * Ensure a potential window is long enough to fit meet request.
+   * @param from     start of window in minutes.
+   * @param end      end of window in minutes.
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:169]: When talking about timeranges, try to be explicit about whether times are inclusive or exclusive (is 10 to 20 a 9, 10, or 11 minute interval?)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,180 @@
...
+     * |-----A-----|                  |------A-----|
+     *       |----------B-----|       |---B---|
+     *
+     * |______________________||______|
+     *        block              window
+     *
+     * We iterate shifts to identify the right-most extent of a block.
+     * If we find a shift that start *after* this point, the difference is a window.
+     * That window is a possible meeting spot, if long enough.
+     */
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:63]: I like this!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,180 @@
...
+    boolean hasMandatoryEvents = false;
+    boolean hasOptionalEvents = false;
+
+    // Whether we should consider attendees of a given class.
+    // This is used to handle cases like all optional no mandatory, no attendees, etc.
+    boolean considerMandatoryAttendees = request.getAttendees().size() > 0;
+    boolean considerOptionalAttendees = request.getOptionalAttendees().size() > 0;
+
+    for (Event event : sortedEvents) {
+      TimeRange when = event.getWhen();
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: Nit: when could be better named.  Once you get to where it's being used, its definition is no longer visible.  eventTimeRange, for example, would be more explicit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,180 @@
...
+     *
+     * |______________________||______|
+     *        block              window
+     *
+     * We iterate shifts to identify the right-most extent of a block.
+     * If we find a shift that start *after* this point, the difference is a window.
+     * That window is a possible meeting spot, if long enough.
+     */
+    int mandatoryBlockExtent = TimeRange.START_OF_DAY; // Based on people who *must* attend
+    int optionalBlockExtent = TimeRange.START_OF_DAY; // Based on people who *may* attend
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:65]: Another question about naming.  Extent implies duration to me.  Should this just be "mandatoryBlockEnd"?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,180 @@
...
+  /** Used to sort events by start time. */
+  public static final Comparator<Event> EVENT_COMPARATOR = new Comparator<Event>() {
+
+    @Override
+    public int compare(Event a, Event b) {
+      return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
+    }
+  };
+
+  /** Finds potential meeting times */
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:39]: Probably worth adding full javadoc here :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,180 @@
...
+    int optionalBlockExtent = TimeRange.START_OF_DAY; // Based on people who *may* attend
+
+    // Whether any events exist for mandatory and optional attendees respectively.
+    boolean hasMandatoryEvents = false;
+    boolean hasOptionalEvents = false;
+
+    // Whether we should consider attendees of a given class.
+    // This is used to handle cases like all optional no mandatory, no attendees, etc.
+    boolean considerMandatoryAttendees = request.getAttendees().size() > 0;
+    boolean considerOptionalAttendees = request.getOptionalAttendees().size() > 0;
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:74]: Covering the case where there are no attendees would be easier if you just add a check to the top of the function.  The function is complex enough that you have to read the full code (or trust the tests) to convince yourself that the correct behaviour is performed in that edge case.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,180 @@
...
+      // As mentioned above, we track 'mandatory' people and 'optional' people as two classes.
+      // For a given class, a window exists if there is time between current event involving that class, 
+      // And the end of the last event involving that class (_____BlockExtent)
+      boolean optionalWindowExists = when.start() > optionalBlockExtent;
+      boolean mandatoryWindowExists = when.start() > mandatoryBlockExtent;
+
+      // Makes sure the requested meeting length fits within the window.
+      boolean optionalWindowLongEnough = optionalWindowExists && windowIsLongEnough(optionalBlockExtent, when.start(), request);
+      boolean mandatoryWindowLongEnough = mandatoryWindowExists && windowIsLongEnough(mandatoryBlockExtent, when.start(), request);
+
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:105]: While this isn't the typical duplicated code, you're writing the same code over and over for the optional and mandatory cases.

No need to change your code, but the thought that comes to me is that instead of doing twice the work in one iteration through the events, you could do two iterations: once for all attendees, and once for mandatory attendees if the first iteration failed to yield results.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (jamohile) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Agreed, changed the names.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (jamohile) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for the idea! Managed to modify this.
